### PR TITLE
soupault: 4.4.0 → 4.5.0

### PR DESCRIPTION
**Pull request by toastal**
I tried `ocaml-ng.ocamlPackages_5_0` but it doesn’t seem Nixpkgs is quite ready for it.

CHANGELOG: https://codeberg.org/PataphysicalSociety/soupault/src/tag/4.5.0/CHANGELOG.md

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by toastal** [comment]
@vbgl Is this too hacky? I had hard needs of camomile 2.0 and lambdasoup 1.0

**Comment by toastal** [comment]
(unsure why it's building locally but not on borg)

File: pkgs/tools/typesetting/soupault/camomile-2_0_0.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# There was little interest in upstreaming this package as a whole, at
+# least for now. See Microsoft GitHub Issue:
+# https://github.com/NixOS/nixpkgs/issues/224318
+
+let
+  pname = "camomile";
+  version = "2.0.0";
+in
+ocamlPackages.buildDunePackage {
+  inherit pname version;
```
Comment (wegank) [inline pkgs/tools/typesetting/soupault/camomile-2_0_0.nix:14]:
```suggestion
ocamlPackages.buildDunePackage rec {
  pname = "camomile";
  version = "2.0.0";
```

**Comment by wegank** [comment]
> (unsure why it's building locally but not on borg)

https://github.com/NixOS/nixpkgs/pull/225735#discussion_r1162800295

File: pkgs/tools/typesetting/soupault/default.nix [hunk 2]
```diff
@@ -22,19 +21,30 @@ ocamlPackages.buildDunePackage {
...
+    (lambdasoup.overrideAttrs (old: rec {
+      version = "1.0.0";
+      buildInputs = (old.buildInputs or [ ]) ++ (with ocamlPackages; [
+        camlp-streams
+      ]);
+      src = builtins.fetchTarball {
+        url = "https://github.com/aantron/lambdasoup/archive/refs/tags/${version}.tar.gz";
+        sha256 = "04jd0rjjqcrzm0w490v2bg0bsyps9qgaw63q0fybp474kcvj369x";
+      };
+    }))
```
Comment (wegank) [inline pkgs/tools/typesetting/soupault/default.nix:47]: Same here. Why not update the original package?

File: pkgs/tools/typesetting/soupault/camomile-2_0_0.nix (hunk 1 above)
Comment (toastal) [reply pkgs/tools/typesetting/soupault/camomile-2_0_0.nix:15]: Aren't there consequences to `rec` + `override*` functions? I've been bitten in the past by it like https://stackoverflow.com/questions/56558483/nix-override-docheck-in-pybind11

File: pkgs/tools/typesetting/soupault/camomile-2_0_0.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  };
+
+  buildInputs = with ocamlPackages; [
+    camlp-streams
+    dune-site
+  ];
+
+  meta = {
+    license = lib.licenses.lgpl21;
+  };
```
[the remark this answers is not in the document]
Comment (toastal) [reply pkgs/tools/typesetting/soupault/camomile-2_0_0.nix:33]: I'm not maintaining it and this is temporary. But the license may be needed for auditing.

**Comment by Alizter** [comment]
In camamolile 2.0.1 there is a comomile_compat lib with compatability bindings, however any libs using cam 1 would have to have their dune files patched. I don't know if it is worth having 2 packages or biting the bullet and patching every lib relying in camo 1 and bumping to camo2.

**Comment by wegank** [comment]
> In camamolile 2.0.1 there is a comomile_compat lib with compatability bindings, however any libs using cam 1 would have to have their dune files patched. I don't know if it is worth having 2 packages or biting the bullet and patching every lib relying in camo 1 and bumping to camo2.

We even have `camomile_0_8_2` in nixpkgs, so aliasing the original one to `camomile_1` isn't so bad.

**Comment by toastal** [comment]
I'm not sure if the OCaml community is fully-sold on migrating at this point to that specific project even if there's an obvious need to get things over to 5.x

File: pkgs/tools/typesetting/soupault/camomile-2_0_0.nix (hunk 1 above)
Comment (toastal) [reply pkgs/tools/typesetting/soupault/camomile-2_0_0.nix:15]: 👍‍

File: pkgs/tools/typesetting/soupault/default.nix [hunk 4]
```diff
@@ -22,19 +20,32 @@ ocamlPackages.buildDunePackage {
...
+    (import ./camomile-2_0_0.nix { inherit lib ocamlPackages fetchFromGitHub; })
     containers
     digestif
+    dune-site
     ezjsonm
     fileutils
     fmt
     jingoo
-    lambdasoup
+    (lambdasoup.overrideAttrs (old: rec {
```
Comment (vbgl) [inline pkgs/tools/typesetting/soupault/default.nix:37]: This local override is not needed, see: #230755.

File: pkgs/tools/typesetting/soupault/default.nix [hunk 5]
```diff
@@ -22,19 +20,32 @@ ocamlPackages.buildDunePackage {
...
     rev = version;
-    sha256 = "sha256-M4gaPxBxQ1Bk2C3BwvobYHyaWKIZgQ6buZ6S5wBlvPg=";
+    sha256 = "sha256-LrDP3IX+UARUN98O8yP7lbyWGg2JT7+MtJumW00IiYw=";
   };
 
   buildInputs = with ocamlPackages; [
     base64
-    camomile
+    camlp-streams
+    (import ./camomile-2_0_0.nix { inherit lib ocamlPackages fetchFromGitHub; })
```
Comment (SuperSandro2000) [inline pkgs/tools/typesetting/soupault/default.nix:29]:
```suggestion
    (callPackage./camomile-2_0_0.nix { })
```

File: pkgs/tools/typesetting/soupault/camomile-2_0_0.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/tools/typesetting/soupault/camomile-2_0_0.nix:30]: meta must be completed with at least license and maintainers and better also with description and homepage.

File: pkgs/tools/typesetting/soupault/default.nix (hunk 5 above)
Comment (toastal) [reply pkgs/tools/typesetting/soupault/default.nix:29]: good call

**Comment by toastal** [comment]
Merge request for the newer version that addresses issues here: https://github.com/NixOS/nixpkgs/pull/238422

**Pull request closed without merging**